src/mlpack/methods/linear_svm/linear_svm_main.cpp

```diff
@@ -0,0 +1,419 @@
+
+  // Epochs needs to be non-negative.
+  RequireParamValue<int>("epochs", [](int x) { return x >= 0; }, true,
+      "maximum number of epochs must be non-negative");
+
+  if (optimizerType != "psgd")
+  {
+    if (CLI::HasParam("step_size"))
+    {
+      Log::Warn << PRINT_PARAM_STRING("step_size") << " ignored because "
+          << "optimizer type is not 'psgd'." << std::endl;
+    }
+    if (CLI::HasParam("shuffle"))
+    {
+      Log::Warn << PRINT_PARAM_STRING("shuffle") << " ignored because "
+          << "optimizer type is not 'psgd'." << std::endl;
+    }
+    if (CLI::HasParam("epochs"))
+    {
+      Log::Warn << PRINT_PARAM_STRING("epochs") << " ignored because "
+          << "optimizer type is not 'psgd'." << std::endl;
+    }
+  }
+
+  // Step Size must be positive.
+  RequireParamValue<double>("step_size", [](double x) { return x >= 0.0; },
+      true, "step size must be positive");
+
+  // Lambda must be positive.
+  RequireParamValue<double>("lambda", [](double x) { return x >= 0.0; },
+      true, "lambda must be positive or zero");
+
+  // Number of Classes must be Non-Negative
+  RequireParamValue<int>("number_of_classes", [](int x) { return x >= 0; },
+                         true, "number of classes must be greater than or "
+                         "equal to 0 (equal to 0 in case of unspecified.)");
+
+  // Delta must be positive.
+  RequireParamValue<double>("delta", [](double x) { return x >= 0.0; }, true,
+      "delta must be non-negative");
+
+  // Delta must be positive.
+  RequireParamValue<int>("epochs", [](int x) { return x > 0; }, true,
+      "epochs must be non-negative");
+
+  // These are the matrices we might use.
+  arma::mat trainingSet;
+  arma::Row<size_t> labels;
+  arma::Row<size_t> rawLabels;
+  arma::mat testSet;
+  arma::Row<size_t> predictions;
+  arma::Col<size_t> mappings;
+  size_t numClasses;
+
+  // Load data matrix.
+  if (CLI::HasParam("training"))
+    trainingSet = std::move(CLI::GetParam<arma::mat>("training"));
+
+  // Check if the labels are in a separate file.
+  if (CLI::HasParam("training") && CLI::HasParam("labels"))
+  {
+    rawLabels = std::move(CLI::GetParam<arma::Row<size_t>>("labels"));
+    if (trainingSet.n_cols != rawLabels.n_cols)
+    {
+      Log::Fatal << "The labels must have the same number of points as the "
+          << "training dataset." << endl;
+    }
+  }
+  else if (CLI::HasParam("training"))
+  {
+    // Checking the size of training data if no labels are passed.
+    if (trainingSet.n_rows < 2)
+    {
+      Log::Fatal << "Can't get labels from training data since it has less "
+          << "than 2 rows." << endl;
+    }
+
+    // The initial predictors for y, Nx1.
+    rawLabels = arma::conv_to<arma::Row<size_t>>::from(
+        trainingSet.row(trainingSet.n_rows - 1));
+    trainingSet.shed_row(trainingSet.n_rows - 1);
+  }
+
+  data::NormalizeLabels(rawLabels, labels, mappings);
+
+  // Load the model, if necessary.
+  LinearSVM<>* model;
```

Comment:
The use of `NormalizeLabels` makes this a little bit tricky.  Consider the following use of the program:

1) a user trains a model passing in data with labels `1, 2, 3` (these will be turned into `0, 1, 2` by `NormalizeLabels()`)
2) now a user saves the model and uses it in a separate call to get some predictions...

but those predictions have labels `0, 1, 2`, not `1, 2, 3`, so they don't match the original labels.  In addition, if the user uses a test set and `test_labels`, the test labels are not mapped in the same way.

So, I think it may be necessary to create a `LinearSVMModel` that also holds the mappings, and then applies the mappings from any saved model to any labels (input or output).

Let me know if I can clarify what I mean. :+1:

Reply:
I have added a `LinearSVMModel` please take a look.